contracts/BananaAllocator.sol

```diff
@@ -0,0 +1,235 @@
+
+            emit AllocationWithdrawn(aid, amount, usdAmount, msg.sender);
+        }
+
+        if (lastSyncTimestamp + syncDelay < block.timestamp) {
+            syncAllocations();
+        }
+    }
+
+    /// @notice An internal function to sync all the allocation amounts
+    function syncAllocations() internal {
+        uint256 length = allocationInfo.length;
+        uint256 totalTokensOwed = 0;
+        uint256 tokensInContract = allocationToken.balanceOf(address(this));
+
+        for (uint256 aid = 0; aid < length; aid++) {
+            totalTokensOwed = totalTokensOwed.add(allocationInfo[aid].tokensAvailable);
+        }
+
+        uint256 tokensToAllocate = tokensInContract.sub(totalTokensOwed);
+        
+        for (uint256 i = 0; i < length; i++) {
+            allocationInfo[i].tokensAvailable += allocationInfo[i].allocationAmount.mul(tokensToAllocate).div(totalAllocations);
+        }
+
+        lastSyncTimestamp = block.timestamp;
+        
+        emit AllocationsSynced();
+    }
+
+    /// @notice A function to adjust the sync delay
+    /// @param _syncDelay number to adjust the sync delay to
+    ///     in milliseconds
+    function setSyncDelay(uint256 _syncDelay) external onlyOwner {
+        syncDelay = _syncDelay;
+
+        emit SyncDelaySet(_syncDelay);
+    }
+
+    /// @notice A function to add a new allocation
+    /// @param allocationName string referencing the name of an allocation
+    /// @param allocationAmount amount for allocation to recieve
+    /// @param newAllocationAdmins array of addresses to be admins in this specific allocation
+    function addAllocation(
+        string memory allocationName, 
+        uint256 allocationAmount, 
+        address[] memory newAllocationAdmins
+    ) external onlyOwner {
+        totalAllocations += allocationAmount;
+
+        allocationInfo.push(AllocationInfo({
+            allocationName: allocationName,
+            allocationAmount: allocationAmount, 
+            tokensAvailable: 0
+        }));
+
+        uint256 _aid = allocationInfo.length - 1;
+
+        for(uint256 i = 0; i < newAllocationAdmins.length; i++) {
+            allocationInfo[_aid].allocationAdmins[newAllocationAdmins[i]] = true;
+        }
+    
+        emit AllocationAdded(allocationName, allocationAmount, newAllocationAdmins);
+    }
+
+    /// @notice An external function to adjust the amount an allocation has
+    /// @param aid number referencing the applicable allocation
+    /// @param allocationAmount the updated amount to set an allocation to
+    function setAllocation(
+        uint256 aid, 
+        uint256 allocationAmount
+    ) external onlyOwner {
+        uint256 prevAllocationAmount = allocationInfo[aid].allocationAmount;
+
+        allocationInfo[aid].allocationAmount = allocationAmount;
+        if (prevAllocationAmount != allocationAmount) {
+            totalAllocations = totalAllocations.sub(prevAllocationAmount).add(allocationAmount);
```

Comment:
It doesn't seem possible for this to fail, but it's a good idea to do `.add` first incase the `.sub` creates a negative value which would cause the tx to fail.